File: src/algorithms/network/odd-even-merge-sort.js
// src/algorithms/network/odd-even-merge.js

import Algorithm from '../core/algorithm-base';

/**
 * Implementation of Odd-Even Merge Sort (Batcher's Merge Exchange Sort) algorithm.
 * 
 * Odd-Even Merge Sort is a sorting network algorithm that builds a network of
 * compare-and-swap operations in a recursive divide-and-conquer approach. It sorts
 * a list by recursively sorting its odd and even indexed sublists, then merging
 * them using an odd-even merging network.
 * 
 * This implementation includes:
 * - Complete visualization of the sorting network structure
 * - Detailed tracking of each compare-and-swap operation
 * - Recursive network construction with explainable steps
 * - Optimal network generation for educational purposes
 * 
 * Time Complexity:
 * - Sequential execution: O(n log² n) for all cases
 * - Parallel execution: O(log² n) with O(n log² n) processors
 * 
 * Space Complexity: O(n) for network representation, O(1) for execution
 * 
 * Stability: Not stable - compare-and-swap operations do not preserve order of equal elements
 * 
 * Parallelization: Highly parallelizable with predictable execution path
 * 
 * @class OddEvenMergeSort
 * @extends Algorithm
 */
class OddEvenMergeSort extends Algorithm {
  /**
   * Create a new OddEvenMergeSort instance
   * 
   * @param {Object} options - Configuration options
   * @param {boolean} [options.visualizeNetwork=true] - Visualize the sorting network
   * @param {boolean} [options.optimizeNetworkSize=true] - Generate optimal network
   * @param {boolean} [options.simulateParallel=false] - Simulate parallel execution
   */
  constructor(options = {}) {
    super('Odd-Even Merge Sort', 'network', options);
    
    // Default options
    this.options = {
      visualizeNetwork: true,      // Visualize the sorting network
      optimizeNetworkSize: true,   // Generate an optimal network
      simulateParallel: false,     // Simulate parallel execution
      ...options
    };
    
    // Sorting network representation
    this.network = [];
    this.executionSteps = [];
  }
  
  /**
   * Execute Odd-Even Merge Sort on the input array
   * 
   * @param {Array} array - Input array to be sorted
   * @param {Object} options - Runtime options
   * @returns {Array} - Sorted array
   */
  run(array, options) {
    // Clone array to avoid modifying the original
    const result = [...array];
    const n = result.length;
    
    // Ensure the array size is a power of 2 by padding
    const padded = this.padToPowerOfTwo(result);
    const paddedLength = padded.length;
    
    this.setPhase('network-generation');
    
    // Generate the sorting network
    this.generateSortingNetwork(paddedLength);
    
    // Record the generated network
    if (this.options.visualizeNetwork) {
      this.recordState(padded, {
        type: 'network-generated',
        network: this.network,
        message: `Generated sorting network with ${this.network.length} comparators`
      });
    }
    
    this.setPhase('sorting');
    
    // Execute the sorting network
    this.executeSortingNetwork(padded);
    
    // Extract the sorted result (removing padding if added)
    const sortedResult = padded.slice(0, n);
    
    this.setPhase('completed');
    return sortedResult;
  }
  
  /**
   * Pad the array to the next power of 2 in length
   * 
   * @param {Array} array - The input array
   * @returns {Array} - Padded array with length that is a power of 2
   * @private
   */
  padToPowerOfTwo(array) {
    const n = array.length;
    
    // If array length is already a power of 2, return a copy
    if ((n & (n - 1)) === 0) {
      return [...array];
    }
    
    // Calculate next power of 2
    const nextPowerOfTwo = Math.pow(2, Math.ceil(Math.log2(n)));
    
    // Create a new array with original elements and padding
    const padded = [...array];
    
    // Pad with the largest value to ensure they go to the end
    // Find the maximum value in the array
    const maxValue = Math.max(...array.filter(x => typeof x === 'number'), Number.MIN_SAFE_INTEGER);
    
    // For non-numeric arrays, use the largest element for padding
    const paddingValue = (typeof maxValue === 'number' && isFinite(maxValue)) ? 
                        Number.MAX_SAFE_INTEGER : 
                        array.reduce((a, b) => (this.compare(a, b) > 0 ? a : b), array[0]);
    
    // Pad the array
    for (let i = n; i < nextPowerOfTwo; i++) {
      padded.push(paddingValue);
    }
    
    // Record the padding operation
    this.recordState(padded, {
      type: 'array-padding',
      originalLength: n,
      paddedLength: nextPowerOfTwo,
      paddingValue: paddingValue,
      message: `Padded array from length ${n} to ${nextPowerOfTwo} with padding value ${paddingValue}`
    });
    
    return padded;
  }
  
  /**
   * Generate the complete sorting network for Odd-Even Merge Sort
   * 
   * @param {number} n - Size of the array
   * @private
   */
  generateSortingNetwork(n) {
    this.network = [];
    this.executionSteps = [];
    
    // Generate the network recursively
    this.oddEvenMergeSort(0, n - 1);
    
    // Optimize the network if enabled
    if (this.options.optimizeNetworkSize) {
      this.optimizeNetwork();
    }
  }
  
  /**
   * Recursive implementation of Odd-Even Merge Sort network generation
   * 
   * @param {number} low - Starting index
   * @param {number} high - Ending index
   * @private
   */
  oddEvenMergeSort(low, high) {
    if (high <= low) {
      return;
    }
    
    // Middle point
    const mid = Math.floor((low + high) / 2);
    
    // Record the division step
    this.executionSteps.push({
      type: 'division',
      ranges: [[low, mid], [mid + 1, high]],
      message: `Dividing range [${low}...${high}] into [${low}...${mid}] and [${mid + 1}...${high}]`
    });
    
    // Sort the two halves recursively
    this.oddEvenMergeSort(low, mid);
    this.oddEvenMergeSort(mid + 1, high);
    
    // Merge the sorted halves
    this.oddEvenMerge(low, high, 1);
  }
  
  /**
   * Odd-Even merge operation for network generation
   * 
   * @param {number} low - Starting index
   * @param {number} high - Ending index
   * @param {number} r - Stride for comparisons
   * @private
   */
  oddEvenMerge(low, high, r) {
    const step = r * 2;
    
    if (step < high - low + 1) {
      // Record the merge step
      this.executionSteps.push({
        type: 'merge',
        range: [low, high],
        stride: r,
        message: `Merging elements in range [${low}...${high}] with stride ${r}`
      });
      
      // Recursively merge even and odd sublists
      this.oddEvenMerge(low, high, step);
      this.oddEvenMerge(low + r, high, step);
      
      // Compare-and-swap elements at stride distance
      for (let i = low + r; i < high; i += step) {
        this.addComparator(i, i + r);
      }
    } else {
      // Base case: Just add a single comparator
      if (r === 1 && low < high) {
        this.addComparator(low, high);
      }
    }
  }
  
  /**
   * Add a comparator to the network
   * 
   * @param {number} i - First index
   * @param {number} j - Second index
   * @private
   */
  addComparator(i, j) {
    // Ensure i < j
    if (i > j) {
      [i, j] = [j, i];
    }
    
    // Add comparator to network
    this.network.push([i, j]);
    
    // Record the comparator addition
    this.executionSteps.push({
      type: 'add-comparator',
      indices: [i, j],
      message: `Adding comparator between indices ${i} and ${j}`
    });
  }
  
  /**
   * Optimize the network by removing redundant comparators
   * 
   * @private
   */
  optimizeNetwork() {
    // This is a simplified optimization - in practice more sophisticated techniques exist
    const originalSize = this.network.length;
    
    // Remove duplicate comparators
    const uniqueComparators = new Set();
    this.network = this.network.filter(([i, j]) => {
      const key = `${i},${j}`;
      if (uniqueComparators.has(key)) {
        return false;
      }
      uniqueComparators.add(key);
      return true;
    });
    
    // Record the optimization step
    this.executionSteps.push({
      type: 'optimize-network',
      originalSize: originalSize,
      optimizedSize: this.network.length,
      reduction: originalSize - this.network.length,
      message: `Optimized network from ${originalSize} to ${this.network.length} comparators`
    });
  }
  
  /**
   * Execute the sorting network on the input array
   * 
   * @param {Array} array - Array to be sorted
   * @private
   */
  executeSortingNetwork(array) {
    // In parallel simulation, we execute stages of the network
    if (this.options.simulateParallel) {
      this.executeParallelNetwork(array);
    } else {
      this.executeSequentialNetwork(array);
    }
  }
  
  /**
   * Execute the network sequentially
   * 
   * @param {Array} array - Array to be sorted
   * @private
   */
  executeSequentialNetwork(array) {
    // Execute each comparator sequentially
    for (let comparatorIdx = 0; comparatorIdx < this.network.length; comparatorIdx++) {
      const [i, j] = this.network[comparatorIdx];
      
      // Record the comparison operation
      this.recordState(array, {
        type: 'comparator',
        indices: [i, j],
        comparatorIndex: comparatorIdx,
        values: [array[i], array[j]],
        message: `Comparing elements at indices ${i} and ${j}`
      });
      
      // Perform compare-and-swap
      if (this.compare(array[i], array[j]) > 0) {
        this.swap(array, i, j);
        
        // Record the swap operation
        this.recordState(array, {
          type: 'network-swap',
          indices: [i, j],
          comparatorIndex: comparatorIdx,
          message: `Swapped elements at indices ${i} and ${j}`
        });
      }
    }
  }
  
  /**
   * Execute the network in parallel stages
   * 
   * @param {Array} array - Array to be sorted
   * @private
   */
  executeParallelNetwork(array) {
    // Identify parallelizable stages in the network
    const stages = this.identifyParallelStages();
    
    // Execute each stage
    for (let stageIdx = 0; stageIdx < stages.length; stageIdx++) {
      const stage = stages[stageIdx];
      
      // Record the stage beginning
      this.recordState(array, {
        type: 'parallel-stage-begin',
        stage: stageIdx + 1,
        comparators: stage,
        message: `Beginning parallel stage ${stageIdx + 1} with ${stage.length} comparators`
      });
      
      // Collect all swaps to perform in this stage
      const swapsToPerform = [];
      
      // Evaluate all comparators in this stage
      for (const [i, j] of stage) {
        // Record the comparison
        this.compare(array[i], array[j]);
        
        // If swap needed, record it
        if (array[i] > array[j]) {
          swapsToPerform.push([i, j]);
        }
      }
      
      // Perform all swaps for this stage in parallel
      if (swapsToPerform.length > 0) {
        // Create a copy for visualization
        const beforeSwap = [...array];
        
        // Apply all swaps
        for (const [i, j] of swapsToPerform) {
          this.swap(array, i, j);
        }
        
        // Record the parallel swap operation
        this.recordState(array, {
          type: 'parallel-stage-swap',
          stage: stageIdx + 1,
          swaps: swapsToPerform,
          beforeState: beforeSwap,
          message: `Performed ${swapsToPerform.length} swaps in parallel during stage ${stageIdx + 1}`
        });
      }
      
      // Record the stage completion
      this.recordState(array, {
        type: 'parallel-stage-complete',
        stage: stageIdx + 1,
        message: `Completed parallel stage ${stageIdx + 1}`
      });
    }
  }
  
  /**
   * Identify parallelizable stages in the network
   * 
   * @returns {Array} - Array of parallel stages, each containing comparators that can execute in parallel
   * @private
   */
  identifyParallelStages() {
    const stages = [];
    const usedIndices = new Set();
    
    // Create a copy of the network for processing
    const remainingComparators = [...this.network];
    
    // While there are comparators left to process
    while (remainingComparators.length > 0) {
      const currentStage = [];
      usedIndices.clear();
      
      // Find all comparators that can execute in parallel
      for (let i = 0; i < remainingComparators.length; i++) {
        const [a, b] = remainingComparators[i];
        
        // If neither index is used in this stage, add to current stage
        if (!usedIndices.has(a) && !usedIndices.has(b)) {
          currentStage.push(remainingComparators[i]);
          usedIndices.add(a);
          usedIndices.add(b);
          
          // Remove from remaining comparators (adjust i to account for removal)
          remainingComparators.splice(i, 1);
          i--;
        }
      }
      
      // Add the current stage
      if (currentStage.length > 0) {
        stages.push(currentStage);
      }
    }
    
    return stages;
  }
  
  /**
   * Get the time and space complexity of Odd-Even Merge Sort
   * 
   * @returns {Object} - Complexity information
   */
  getComplexity() {
    return {
      time: {
        best: 'O(n log² n)',
        average: 'O(n log² n)',
        worst: 'O(n log² n)',
        parallel: 'O(log² n) with O(n log² n) processors'
      },
      space: {
        best: 'O(n)',   // For network representation
        average: 'O(n)',
        worst: 'O(n)'
      },
      comparators: 'O(n log² n)'  // Network size
    };
  }
  
  /**
   * Whether Odd-Even Merge Sort is stable
   * 
   * @returns {boolean} - False as Odd-Even Merge Sort is not stable
   */
  isStable() {
    return false;
  }
  
  /**
   * Whether Odd-Even Merge Sort is in-place
   * 
   * @returns {boolean} - True for execution, false considering network storage
   */
  isInPlace() {
    return true;  // The sorting itself is in-place, though network generation is not
  }
  
  /**
   * Get detailed algorithm information
   * 
   * @returns {Object} - Detailed algorithm metadata
   */
  getInfo() {
    const info = super.getInfo();
    
    // Add Odd-Even Merge Sort specific information
    info.optimization = {
      visualizeNetwork: this.options.visualizeNetwork,
      optimizeNetworkSize: this.options.optimizeNetworkSize,
      simulateParallel: this.options.simulateParallel
    };
    
    info.properties = {
      isNetwork: true,
      comparisonBased: true,
      stable: false,
      inPlace: true,  // Execution is in-place
      dataOblivious: true,  // Performance is independent of input data
      parallelizable: true
    };
    
    info.networkCharacteristics = {
      size: 'O(n log² n) comparators',
      depth: 'O(log² n) parallel steps',
      regularity: 'Highly regular structure',
      modularity: 'Composed of merge networks',
      verification: 'Provably correct by construction'
    };
    
    info.parallelCharacteristics = {
      granularity: 'Fine-grained',
      synchronization: 'Barrier synchronization between stages',
      communicationPattern: 'Local communication between adjacent processors',
      scalability: 'Excellent with O(n log² n) processors',
      speedup: 'O(n) speedup with sufficient processors'
    };
    
    info.applications = [
      'Hardware sorting networks',
      'FPGA implementations',
      'Systolic arrays',
      'Parallel computing environments',
      'High-throughput sorting applications'
    ];
    
    info.advantages = [
      'Data-oblivious with consistent performance regardless of input',
      'Highly parallelizable with predictable execution path',
      'Well-suited for hardware implementation',
      'No conditional branches or complex data structures',
      'Theoretically elegant with mathematical properties'
    ];
    
    info.disadvantages = [
      'Not competitive sequentially with O(n log n) algorithms',
      'Not adaptive to pre-sorted or partially sorted inputs',
      'Not stable - does not preserve order of equal elements',
      'Network size grows as O(n log² n) rather than O(n log n)',
      'Requires padding to power-of-two size for standard implementation'
    ];
    
    info.relatedAlgorithms = [
      'Bitonic Sort (another network sorting algorithm)',
      'Batcher\'s Odd-Even Mergesort (an optimization of this algorithm)',
      'Pairwise Sorting Networks',
      'Merge Sort (shares divide-and-conquer approach)',
      'Shellsort (also uses incremental sequence of comparisons)'
    ];
    
    info.theoreticalFoundation = [
      'Recursively divides array into odd-indexed and even-indexed elements',
      'Sorts the two subarrays recursively',
      'Merges using a specialized odd-even merging network',
      'Builds a sorting network with O(n log² n) comparators',
      'Achieves optimal O(log² n) parallel time complexity'
    ];
    
    return info;
  }
}

export default OddEvenMergeSort;

File: python/algorithms/network/odd-even-merge-sort.py
# algorithms/network/odd_even_merge_sort.py

from typing import List, Any, Dict, Optional, Callable, Tuple
from algorithms.base_algorithm import Algorithm
import time
import math

class OddEvenMergeSort(Algorithm):
    """
    Implementation of Odd-Even Merge Sort (Batcher's Merge Exchange Sort).
    
    Odd-Even Merge Sort is a parallel sorting algorithm that can be represented as a sorting
    network. It divides the input into two halves, recursively sorts them, and then performs
    a specialized merge operation by comparing and exchanging elements at specific distances.
    
    This implementation includes:
    - Detailed visualization of the network structure
    - Step-by-step execution of the comparator operations
    - Analysis of sorting network properties
    - Multiple optimization strategies
    
    Time Complexity:
    - Sequential implementation: O(n log²(n))
    - Parallel implementation: O(log²(n)) with O(n log²(n)) processors
    
    Space Complexity: O(1) (in-place sorting)
    
    Note: Although the algorithm is designed for parallel execution, this implementation
    simulates the parallel behavior sequentially while preserving the original operation
    order to correctly visualize the sorting network structure.
    """
    
    def __init__(self, options: Dict[str, Any] = None):
        """
        Initialize Odd-Even Merge Sort with options.
        
        Args:
            options: Dictionary of options including:
                - visualize_network: Visualize the sorting network structure
                - optimize_comparators: Optimize redundant comparator operations
                - adaptive_termination: Early termination if array becomes sorted
        """
        super().__init__("Odd-Even Merge Sort", "network", options)
        
        # Default options
        self.options.update({
            "visualize_network": True,     # Visualize the sorting network structure
            "optimize_comparators": True,  # Optimize redundant comparator operations
            "adaptive_termination": True,  # Early termination if array becomes sorted
            "animation_delay": 0           # Delay between steps for visualization
        })
        
        # Override with provided options
        if options:
            self.options.update(options)
        
        # Keep track of the comparator network structure
        self.network_structure = {
            "stages": [],
            "total_comparators": 0
        }

    def run(self, array: List[Any], options: Dict[str, Any]) -> List[Any]:
        """
        Execute Odd-Even Merge Sort on the provided array.
        
        Args:
            array: The array to sort
            options: Runtime options
            
        Returns:
            The sorted array
        """
        # Make a copy to avoid modifying the original
        result = array.copy()
        n = len(result)
        
        # For Odd-Even Merge Sort, n should ideally be a power of 2
        # If not, we can pad the array with large values that will be filtered out later
        original_length = n
        next_power_of_2 = 2 ** math.ceil(math.log2(n))
        
        if n != next_power_of_2:
            # Record padding operation
            self.record_state(result, {
                "type": "padding",
                "original_length": n,
                "padded_length": next_power_of_2,
                "message": f"Padding array from length {n} to {next_power_of_2} for Odd-Even Merge Sort"
            })
            
            # Find the largest value in the array to use for padding
            max_value = max(result) if result else float('inf')
            # Pad with values larger than any in the array
            padding = [max_value + 1] * (next_power_of_2 - n)
            result.extend(padding)
            n = next_power_of_2
        
        self.set_phase("sorting")
        
        # Reset network structure
        self.network_structure = {
            "stages": [],
            "total_comparators": 0
        }
        
        # Execute the odd-even merge sort
        self.odd_even_merge_sort(result, 0, n, options)
        
        # If we padded the array, remove the padding
        if original_length != next_power_of_2:
            result = result[:original_length]
            
            # Record unpadding operation
            self.record_state(result, {
                "type": "unpadding",
                "message": f"Removing padding to restore original array length of {original_length}"
            })
        
        self.set_phase("completed")
        
        # Record final network statistics
        if self.options["visualize_network"]:
            self.record_state(result, {
                "type": "network_statistics",
                "stages": len(self.network_structure["stages"]),
                "comparators": self.network_structure["total_comparators"],
                "message": f"Sorting network: {len(self.network_structure['stages'])} stages with {self.network_structure['total_comparators']} comparators"
            })
        
        return result

    def odd_even_merge_sort(self, 
                           array: List[Any], 
                           lo: int, 
                           n: int, 
                           options: Dict[str, Any]) -> None:
        """
        Recursively apply Odd-Even Merge Sort to the array.
        
        Args:
            array: The array to sort
            lo: Low index of the section to sort
            n: Size of the section to sort
            options: Runtime options
        """
        # Base case: array with 1 element is already sorted
        if n <= 1:
            return
        
        # Record current recursive call
        self.record_state(array, {
            "type": "recursive_call",
            "section": [lo, lo + n - 1],
            "size": n,
            "message": f"Sorting section from index {lo} to {lo + n - 1} (size {n})"
        })
        
        # Introduce delay for visualization if specified
        if options["animation_delay"] > 0:
            time.sleep(options["animation_delay"] / 1000)
        
        # Divide the array into two halves and sort them recursively
        m = n // 2
        
        # Sort the first half
        self.odd_even_merge_sort(array, lo, m, options)
        
        # Sort the second half
        self.odd_even_merge_sort(array, lo + m, m, options)
        
        # Merge the two halves using odd-even merge
        if n > 1:
            self.record_state(array, {
                "type": "merge_start",
                "first_half": [lo, lo + m - 1],
                "second_half": [lo + m, lo + n - 1],
                "message": f"Merging sections [{lo}...{lo + m - 1}] and [{lo + m}...{lo + n - 1}]"
            })
            
            self.odd_even_merge(array, lo, n, 1, options)
            
            self.record_state(array, {
                "type": "merge_complete",
                "section": [lo, lo + n - 1],
                "message": f"Completed merge of section [{lo}...{lo + n - 1}]"
            })

    def odd_even_merge(self, 
                      array: List[Any], 
                      lo: int, 
                      n: int, 
                      r: int, 
                      options: Dict[str, Any]) -> None:
        """
        Merge operation for Odd-Even Merge Sort.
        
        Args:
            array: The array being sorted
            lo: Low index of the section to merge
            n: Size of the section to merge
            r: Distance between elements to compare
            options: Runtime options
        """
        # Create a new stage for the network visualization
        current_stage = {
            "comparators": [],
            "distance": r
        }
        
        # Adaptive termination check
        if options["adaptive_termination"] and self._is_sorted(array, lo, lo + n - 1):
            self.record_state(array, {
                "type": "early_termination",
                "section": [lo, lo + n - 1],
                "message": f"Section [{lo}...{lo + n - 1}] is already sorted, skipping merge"
            })
            return
        
        # Base case: Compare a single pair of elements
        if n == 2:
            self.compare_and_swap(array, lo, lo + r, current_stage)
            if self.options["visualize_network"] and current_stage["comparators"]:
                self.network_structure["stages"].append(current_stage)
            return
        
        # Recursive case: Divide and merge
        m = 2 * r
        
        # Recursively handle odd-even merge
        self.odd_even_merge(array, lo, n, m, options)
        self.odd_even_merge(array, lo + r, n, m, options)
        
        # Compare and swap elements at distance r
        for i in range(lo + r, lo + n - r, m):
            self.compare_and_swap(array, i, i + r, current_stage)
        
        # Add the stage to the network structure if it contains comparators
        if self.options["visualize_network"] and current_stage["comparators"]:
            self.network_structure["stages"].append(current_stage)
            self.network_structure["total_comparators"] += len(current_stage["comparators"])

    def compare_and_swap(self, 
                        array: List[Any], 
                        i: int, 
                        j: int, 
                        stage: Dict[str, Any]) -> None:
        """
        Compare and swap two elements if they are in the wrong order.
        
        Args:
            array: The array being sorted
            i: First index
            j: Second index
            stage: Current network stage for recording comparators
        """
        # Skip invalid indices
        if i >= len(array) or j >= len(array):
            return
        
        # Skip optimization if elements are already in order
        if self.options["optimize_comparators"]:
            if self.compare(array[i], array[j]) <= 0:
                # Elements are already in order, record comparison but don't swap
                if self.options["visualize_network"]:
                    stage["comparators"].append({
                        "indices": [i, j],
                        "swapped": False
                    })
                return
        
        # Compare elements and swap if needed
        if self.compare(array[i], array[j]) > 0:
            self.swap(array, i, j)
            
            # Record the comparator operation
            if self.options["visualize_network"]:
                stage["comparators"].append({
                    "indices": [i, j],
                    "swapped": True
                })
            
            # Record the state after swap
            self.record_state(array, {
                "type": "comparator_swap",
                "indices": [i, j],
                "distance": j - i,
                "values": [array[i], array[j]],
                "message": f"Swapped elements at indices {i} and {j} (distance {j - i})"
            })
        else:
            # Record the comparator operation (no swap)
            if self.options["visualize_network"]:
                stage["comparators"].append({
                    "indices": [i, j],
                    "swapped": False
                })

    def _is_sorted(self, array: List[Any], start: int, end: int) -> bool:
        """
        Check if a section of the array is already sorted.
        
        Args:
            array: The array to check
            start: Start index
            end: End index
            
        Returns:
            True if the section is sorted, False otherwise
        """
        for i in range(start, end):
            if self.compare(array[i], array[i + 1]) > 0:
                return False
        return True

    def get_complexity(self) -> Dict[str, Dict[str, str]]:
        """
        Get the time and space complexity of Odd-Even Merge Sort.
        
        Returns:
            Complexity information dictionary
        """
        return {
            "time": {
                "best": "O(n log²(n))",
                "average": "O(n log²(n))",
                "worst": "O(n log²(n))"
            },
            "space": {
                "best": "O(1)",
                "average": "O(1)",
                "worst": "O(1)"
            },
            "parallel": {
                "time": "O(log²(n))",
                "processors": "O(n log²(n))"
            }
        }

    def is_stable(self) -> bool:
        """
        Whether Odd-Even Merge Sort is stable (preserves relative order of equal elements).
        
        Returns:
            False as the standard implementation is not stable
        """
        return False

    def is_in_place(self) -> bool:
        """
        Whether Odd-Even Merge Sort is in-place (uses O(1) auxiliary space).
        
        Returns:
            True as Odd-Even Merge Sort is in-place
        """
        return True
    
    def get_info(self) -> Dict[str, Any]:
        """
        Get detailed information about the algorithm.
        
        Returns:
            Dictionary with algorithm details
        """
        info = super().get_info()
        
        # Add Odd-Even Merge Sort specific information
        info.update({
            "optimization": {
                "visualize_network": self.options.get("visualize_network", True),
                "optimize_comparators": self.options.get("optimize_comparators", True),
                "adaptive_termination": self.options.get("adaptive_termination", True)
            },
            "properties": {
                "comparison_based": True,
                "stable": False,
                "in_place": True,
                "parallel": True,
                "sorting_network": True
            },
            "network_properties": {
                "depth": "O(log²(n))",
                "size": "O(n log²(n))",
                "comparators": "O(n log²(n))"
            },
            "applications": [
                "Hardware sorting implementations",
                "Parallel computing environments",
                "FPGA and ASIC sorting circuits",
                "Multi-core CPU and GPU sorting",
                "Distributed systems with fixed communication patterns"
            ],
            "advantages": [
                "Fixed communication pattern ideal for hardware implementation",
                "Highly parallelizable with clear communication topology",
                "Efficient for specialized hardware implementations",
                "Predictable performance regardless of input distribution",
                "Well-suited for vectorization and SIMD operations"
            ],
            "disadvantages": [
                "Less efficient than other algorithms in sequential implementations",
                "Higher complexity than optimal comparison-based sorts",
                "Not adaptive to already-sorted or partially-sorted inputs",
                "More complex implementation compared to standard sorting algorithms",
                "Not stable in its standard implementation"
            ],
            "historical_significance": "Developed by Kenneth E. Batcher in 1968, Odd-Even Merge Sort was one of the first practical parallel sorting algorithms and remains important in hardware implementation of sorting networks."
        })
        
        return info


# Example usage
if __name__ == "__main__":
    # Create and run a simple test
    sorter = OddEvenMergeSort({
        "visualize_network": True,
        "optimize_comparators": True
    })
    
    test_array = [5, 3, 8, 4, 2, 9, 1, 7, 6]
    
    result = sorter.execute(test_array)
    print(f"Original array: {test_array}")
    print(f"Sorted array: {result}")
    print(f"Comparisons: {sorter.metrics['comparisons']}")
    print(f"Swaps: {sorter.metrics['swaps']}")
    
    # Print network statistics
    if sorter.options["visualize_network"]:
        print(f"Network stages: {len(sorter.network_structure['stages'])}")
        print(f"Total comparators: {sorter.network_structure['total_comparators']}")
